src/main/java/com/es/phoneshop/web/CartPageServlet.java

```diff
@@ -0,0 +1,95 @@
+package com.es.phoneshop.web;
+
+import com.es.phoneshop.exceptions.OutOfStockException;
+import com.es.phoneshop.model.cart.Cart;
+import com.es.phoneshop.model.product.Product;
+import com.es.phoneshop.service.CartService;
+import com.es.phoneshop.service.ProductService;
+import com.es.phoneshop.service.impl.CartServiceImpl;
+import com.es.phoneshop.service.impl.ProductServiceImpl;
+import jakarta.servlet.ServletConfig;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.HttpServlet;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+
+import java.io.IOException;
+import java.text.NumberFormat;
+import java.text.ParseException;
+import java.util.HashMap;
+import java.util.Map;
+
+public class CartPageServlet extends HttpServlet {
+
+    private ProductService productService;
+
+    private CartService cartService;
+
+    @Override
+    public void init(ServletConfig config) throws ServletException {
+        super.init(config);
+        productService = ProductServiceImpl.getInstance();
+        cartService = CartServiceImpl.getInstance();
+    }
+
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        Cart cart = cartService.getCart(request);
+        if (cart != null) {
+            request.setAttribute("cart", cart);
+            request.getRequestDispatcher("/WEB-INF/pages/cart.jsp").forward(request, response);
+        }
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        String[] productIds = request.getParameterValues("productId");
+        String[] quantities = request.getParameterValues("quantity");
+        Map<Long, String> errors = new HashMap<>();
+
+        if (productIds != null) {
+            for (int i = 0; i < productIds.length; ++i) {
+                Long productId = Long.valueOf(productIds[i]);
+
+                int quantity;
+                try {
+                    NumberFormat format = NumberFormat.getInstance(request.getLocale());
+                    quantity = format.parse(quantities[i]).intValue();
+                    Product product = productService.getProduct(productId);
+                    cartService.updateCart(product, quantity, request);
+                } catch (ParseException | OutOfStockException ex) {
+                    handleException(errors, productId, ex);
```

Comment:
errorMap can be fulfilled just in catch blocks, without checking the class of exception
just catch them and update your errorMap accordingly